### PR TITLE
Added tapenade checks

### LINKED ISSUE #15: Add Tapenade checks
# Description of feature
For repos that contain AD'd code, we should have checks to make sure PRs that touched Fortran code have consistent AD code. This means that Tapenade should produce no diffs.

# Potential solution
Just run Tapenade in the CI container, and then run `git diff --exit-code` or something so that any diffs get reported as a CI failure.

### PR DESCRIPTION
**Pull request by nwu63**
## Purpose
Closes #15. Supports both 3.10 and 3.16.

## Type of change
What types of change is it?
_Select the appropriate type(s) that describe this PR_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

### CODE REVIEW

**Comment by eirikurj** [comment]
Older versions of tapenade do exist in their gitlab repo https://gitlab.inria.fr/tapenade/tapenade
Cloning the repo and checking out the correct tag to recover the older versions should work. For us, the `tap-3.10-5363` tag should be what we use. 
Should we not try to first integrate the older tapenade version, so we can add tests to the current committed code in our repos?

**Comment by nwu63** [comment]
I tried this locally but could not get the older code to run, I got the error
```
Error: Could not find or load main class topLevel.Tapenade
Caused by: java.lang.ClassNotFoundException: topLevel.Tapenade
```
despite setting $TAPENADE_HOME and $JAVA_HOME. So I suspect that the sources have to be built (which I then tried to do following instructions [here](https://gitlab.inria.fr/tapenade/tapenade/-/tree/dc98a2747515d83b91d957f4ad1364077ae37366/distrib) but did not get that to work, possibly due to incompatible Java versions.

At this point, I feel building from source is a bit too much hassle which is why I opted to just work with the new version. If they are okay with us re-distributing our own 3.10 binaries then we could just use that instead and it should work as done with 3.16 here.

**Comment by eirikurj** [comment]
Since tapenade 3.10 is effectively available, I think it's probably ok to self-host the 3.10 tar and use or our testing.

**Comment by nwu63** [comment]
I have updated the PR with support for 3.10. Since we can't test this there will probably be things to fix once this is merged and we use this on a PR. I will probably just directly push to master to fix that then.

**Review by bernardopacini — CHANGES_REQUESTED** [review]
Overall this looks good to me, but I added a few notes that might be issues when we implement this.

I have two additional questions as well:

1) Should we place this in a file that is separate from the style check? I agree it does not belong with the build or test scripts, but it seems like it may warrant a separate file as it is not exactly a style check.

2) I find that Tapenade uses Docker when running on both my Mac and Ubuntu machines. I am not sure if Tapenade can run without it. If it needs Docker while running, will this be an issue for us?

File: azure/azure_style.yaml [hunk 1]
```diff
@@ -75,3 +81,28 @@ jobs:
...
+      - task: JavaToolInstaller@0
+        inputs:
+          versionSpec: "8"
+          jdkArchitectureOption: "x86"
+          jdkSourceOption: "PreInstalled"
+      - script: |
+          if [[ "${{ parameters.TAPENADE_VERSION }}" == "3.10" ]]; then
+            wget http://umich.edu/~mdolaboratory/misc/tapenade3.10.tar
+          elif [[ "${{ parameters.TAPENADE_VERSION }}" == "3.16" ]]; then
+            wget http://www-sop.inria.fr/ecuador/tapenade/distrib/tapenade_3.16.tar
```
[the remark this answers is not in the document]
Comment (nwu63) [reply azure/azure_style.yaml:100]: I just tried this in Docker and it seems to work, we will have to try this on a PR I think to see.

File: azure/azure_style.yaml [hunk 2]
```diff
@@ -75,3 +81,28 @@ jobs:
...
+    pool:
+      vmImage: "ubuntu-20.04"
+    condition: ${{ parameters.TAPENADE }}
+    steps:
+      - checkout: azure_template
+      - task: JavaToolInstaller@0
+        inputs:
+          versionSpec: "8"
+          jdkArchitectureOption: "x86"
+          jdkSourceOption: "PreInstalled"
```
[the remark this answers is not in the document]
Comment (nwu63) [reply azure/azure_style.yaml:95]: Yep according to the [docs](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/tool/java-tool-installer?view=azure-devops) the environment variable is automatically set.

**Comment by nwu63** [comment]
>     1. Should we place this in a file that is separate from the style check? I agree it does not belong with the build or test scripts, but it seems like it may warrant a separate file as it is not exactly a style check.

Good point, I will move this to a separate check then. Can you remind me what exactly does this do? Obviously the file structure will be different in the `.github` repo but does this affect anything running these checks? Just the heading on the individual jobs or something?

>     2. I find that Tapenade uses Docker when running on both my Mac and Ubuntu machines. I am not sure if Tapenade can run without it. If it needs Docker while running, will this be an issue for us?

Docker is available on all these machines so I think we are okay here, again won't know until we try it.

**Comment by bernardopacini** [comment]
> > ```
> > 1. Should we place this in a file that is separate from the style check? I agree it does not belong with the build or test scripts, but it seems like it may warrant a separate file as it is not exactly a style check.
> > ```
> 
> 
> Good point, I will move this to a separate check then. Can you remind me what exactly does this do? Obviously the file structure will be different in the `.github` repo but does this affect anything running these checks? Just the heading on the individual jobs or something?
> 
> 

Do you mean what exactly making a new file does? It will create a new block on Azure, just like the style and build / test ones. Other than that, it will look just like the others.

Another option is to simply make the "Style" name more general, such as "Code Consistency" or something along those lines and avoid a new file.

**Comment by bernardopacini** [comment]
These updates look good to me. I am not sure if in the long run we have any other checks that we can combine into the same file with Tapenade, in which case the `Azure_Tapenade` file might need a new name, but at least for what we need now this looks good to me.

**Review by bernardopacini — APPROVED** [review]

**Pull request merged**